pkg/compactor/compactor.go

```diff
@@ -670,6 +679,12 @@ func (c *Compactor) compactUsers(ctx context.Context) {
 		level.Info(c.logger).Log("msg", "starting compaction of user blocks", "user", userID)
 
 		if err = c.compactUserWithRetries(ctx, userID); err != nil {
+			if ctx.Err() != nil && ctx.Err() == context.Canceled {
```

Comment:
Wouldn't this mask the `err`?  Don't we still want to record down that there was an `err` in this case?

Reply:
I see. I think it's better to emit metrics such that we can detect the edge case you describe indeed happened (both failure and interrupted metrics were emitted); instead of guessing if it was the edge case when we see the interrupted metrics emitted.

Also, for line 650 above you mentioned:

> It's not really clear whether we would benefit from a stricter check

I a wondering what benefit do we have for stricter check here then?

Reply:
> I think it's better to emit metrics such that we can detect the edge case you describe indeed happened (both failure and interrupted metrics were emitted); instead of guessing if it was the edge case when we see the interrupted metrics emitted.

Agreed that would be better but we don't have a good way to do that without refactoring.  Given the impact of hitting that edge case is a short delay in detecting a deterministic failure or the benefit of a retry in the case of transient errors, I don't see the value in refactoring.

If we emit both metrics, we are not fulfilling the core premise of this PR, which is to not signal compaction failure on restarts.

> I a wondering what benefit do we have for stricter check here then?

The stricter check here isn't just for added benefit, it fulfills the requirement.  A canceled context specifically indicates here that we are shutting down, whereas other context errors do not.

In the case of the other ctx.Err() check, I'm just saying it isn't clear what the original intent of that check is and I'd say it seems out of scope here to investigate and try to reason about whether we should update the existing code.

Reply:
https://github.com/prometheus/prometheus/pull/10772 What about using a similar way here to check the returned err instead of `ctx.Err()`?
If the edge case you meant is `ctx.Err() == ctx.Cancelled` whilte err is actually a compaction failure then this should catch it.

Reply:
`errors.Is` will not work here because of the way that thanos wraps errors during compaction. 

Several of the intermediate errors that wrap failures during compaction like `compact.HaltError` and `errutil.NonNilMultiError` do not implement `.Unwrap()` and therefore do not work with `errors.Is` here.

TL;DR on how `errors.Is` works

`errors.Is(foo, bar)` calls `errors.Unwrap` on foo and its error chain until `errors.Unwrap` returns nil or `errors.Unwrap(x) == bar`.  If an error type does not implement `Unwrap()` then `errors.Unwrap()` returns nil.

Reply:
What I am saying is to use the similar to check err rather than `ctx.Err()`. You can choose `Is` or `Cause` and see which one fits. Thanos also has `IsHaltError` for you to check whether it is halt error. If that's the only exception. And you can always implement `Unwrap` on the Thanos side, right?

Reply:
> What I was saying is to use the similar to check err rather than ctx.Err(). You can choose Is or Cause and see which one fits

I think I responded directly to why this does not work in this case, let me know if I'm missing something.  All the problems with `Is` also exist with `Cause`, i.e. the underlying error types need to explicitly support these methods.

> Thanos also has IsHaltError for you to check whether it is halt error. If that's the only exception

Being able to check if the error is a `compact.HaltError` doesn't do anything for us because we still cannot unwrap HaltError to check whether `context.Canceled` is in the error chain.

> And you can always implement Unwrap on the Thanos side, right?

Sure, but I don't think Thanos updates should block this change, especially given this is already how we check for compactor interruption in the code.  There is precedent here.

Reply:
> I think I responded directly to why this does not work in this case, let me know if I'm missing something. All the problems with Is also exist with Cause, i.e. the underlying error types need to explicitly support these methods.

Anyway we are not checking err directly, we are checking `ctx.Err()` so anyway this doesn't solve the edge case, if this is the thing you and Alvin discussed. So I feel the best way is still to patch Thanos with the right method and check.

> given this is already how we check for compactor interruption in the code. There is precedent here.

If you mean the ctx check above I think it is more a check to handle closed ctx. For the wrapped error it is different.

Reply:
I don't want to block this. This is good as a first step. We can try to patch Thanos in a separate pr and update the code if necessary

Reply:
I agree with you that updating Thanos to allow for better error checking here is ideal, for me it is just about whether or not we are blocking this.  I can log an action item to patch Thanos and come back to this.

Reply:
Yeah, let's not block it!